last_ordered_products/models/product_product.py

```diff
@@ -0,0 +1,132 @@
+from datetime import datetime
+from odoo import api, fields, models
+from odoo.osv import expression
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    last_order_time = fields.Datetime(compute='_compute_last_order_time')
+    last_date_str = fields.Char(compute='_compute_last_order_time')
+
+    @api.depends_context('order_id')
+    def _compute_last_order_time(self):
+        """Compute the last order time for each product based on the latest sale or purchase."""
+
+        order_type = False
+        if self.env.context.get('active_model') == 'sale.order.line':
+            partner_id = self.env['sale.order'].browse(self.env.context.get('order_id')).partner_id.id
+            order_type = 'sale'
+        elif self.env.context.get('active_model') == 'purchase.order.line':
+            partner_id = self.env['purchase.order'].browse(self.env.context.get('order_id')).partner_id.id
+            order_type = 'purchase'
+        elif self.env.context.get('active_model') == 'account.journal':
+            active_id = self.env.context.get('active_id')
+            if active_id:
+                order_type = self.env['account.journal'].browse(active_id).type
+            partner_id = self.env.context.get('partner_id') or self.env.context.get('default_partner_id')
+        else:
+            partner_id = self.env.context.get('partner_id')
+            order_type = self.env.context.get('order_type')
```

Comment:
We should not rely on context too much instead add something robust to compute these fields